File: python/problems/leetcode/0215_kth_largest_element_in_an_array/solution.py
"""LeetCode 215: Kth Largest Element in an Array."""

from __future__ import annotations

import heapq


class Solution:
    def findKthLargest(self, nums: list[int], k: int) -> int:
        heap: list[int] = []
        for x in nums:
            heapq.heappush(heap, x)
            if len(heap) > k:
                heapq.heappop(heap)
        return heap[0]


if __name__ == "__main__":
    assert Solution().findKthLargest([3, 2, 1, 5, 6, 4], 2) == 5
    assert Solution().findKthLargest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4) == 4
    print("0215 OK")

File: cpp/problems/leetcode/0215_kth_largest_element_in_an_array/solution.cpp
// LeetCode 215
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int x : nums) {
            pq.push(x);
            if ((int)pq.size() > k) pq.pop();
        }
        return pq.top();
    }
};

int main() {
    vector<int> a{3, 2, 1, 5, 6, 4};
    assert(Solution().findKthLargest(a, 2) == 5);
    vector<int> b{3, 2, 3, 1, 2, 4, 5, 5, 6};
    assert(Solution().findKthLargest(b, 4) == 4);
    cout << "0215 OK" << endl;
    return 0;
}
